test_runner/regress/test_tenant_detach.py

```diff
@@ -108,6 +108,147 @@ def test_tenant_reattach(
+
+    # Sanity check that the failpoint was reached
+    assert env.pageserver.log_contains('failpoint "wal-ingest-logical-message-sleep": sleep done')
+    assert time.time() - before_attach_time > 5
+
+    # Clean up
+    pageserver_http.configure_failpoints(("wal-ingest-logical-message-sleep", "off"))
+
+
+# Detach and re-attach a tenant with a branch
+#
+# This exercises the same race condition after tenant attach as
+# 'test_tenant_with_branches_reattach_slow_replay', but this test case
+# is closer to the original scenario where we originally found the
+# issue.
+#
+# In this scenario, the incorrect result to get-request leads to
+# *permanent damage* in the child timeline, because ingesting the WAL
+# on the child timeline depended on incorrect view of the parent. This
+# test reproduced one such case; the symptom was a 'FATAL: database
+# "neondb" does not exist' error on the child, when trying to connect
+# after re-attaching the tenant. The WAL record of CREATE DATABASE was
+# missed when searching the DbDirectory key in the ancestor.
+@pytest.mark.parametrize("remote_storage_kind", available_remote_storages())
+def test_tenant_with_branches_reattach_createdb(
+    neon_env_builder: NeonEnvBuilder,
+    remote_storage_kind: RemoteStorageKind,
+):
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_tenant_with_branches_reattach_createdb",
+    )
+
+    env = neon_env_builder.init_start()
+    pageserver_http = env.pageserver.http_client()
+
+    # create new nenant
+    tenant_id, timeline_id = env.neon_cli.create_tenant()
+
+    main_endpoint = env.endpoints.create_start("main", tenant_id=tenant_id)
+    with main_endpoint.cursor() as cur:
+        cur.execute("SELECT pg_logical_emit_message(false, 'neon-test', 'between inserts')")
+
+        cur.execute("CREATE DATABASE neondb")
+        current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
+    env.neon_cli.create_branch("child", tenant_id=tenant_id, ancestor_start_lsn=current_lsn)
+
+    with main_endpoint.cursor(dbname="neondb") as cur:
+        cur.execute("CREATE TABLE test_migrate_one AS SELECT generate_series(1,100)")
+    main_endpoint.stop()
+
+    child_endpoint = env.endpoints.create_start("child", tenant_id=tenant_id)
+    with child_endpoint.cursor(dbname="neondb") as cur:
+        cur.execute("CREATE TABLE test_migrate_one AS SELECT generate_series(1,200)")
+    child_endpoint.stop()
+
+    pageserver_http.tenant_detach(tenant_id)
+
+    # Wait before ingesting the WAL for CREATE DATABASE on the main branch. The original
+    # bug reproduced easily even without this, as there is always some delay between
+    # loading the timeline and establishing the connection to the safekeeper to stream and
+    # ingest the WAL, but let's make this less dependent on accidental timing.
+    pageserver_http.configure_failpoints([("wal-ingest-logical-message-sleep", "return(5000)")])
```

Comment:
The `Wait before ingesting the WAL for CREATE DATABASE` is a lie, no?
It's just that, usually, the Python script is faster than the `CREATE DATABASE` WAL reaching pageservers.
Or am I missing something?

Also, for my understanding: before the fixes of this patch, what happened, exactly?
We create the child branch at the right LSN, but, didn't have the data.
So, the child branch asked safekeepers for the WAL of _what_ timeline? the child one or the parent one? And how did they respond?

Reply:
Added a more detailed explanation. Let me know if it makes it clear or not